languages/tree-sitter-stack-graphs-java/src/stack-graphs.tsg

```diff
@@ -259,60 +560,266 @@ attribute node_symbol = node            => symbol = (source-text node), source_n
+
+  node def
+  attr (def) node_definition = @id
+
+  node def__typeof
+  attr (def__typeof) pop_symbol = ":"
+
+  edge def -> def__typeof
+
+  edge def__typeof -> @type.type
+
+  edge @this.after_scope -> def
+  edge @this.after_scope -> @this.before_scope
+}
+
+(resource . (identifier) @id .) @this {
+  node @id.lexical_scope
+  node @id.type
+
+  node ref
+  node ref__typeof
+
+  attr (ref) node_reference = @id
+
+  edge ref -> @id.lexical_scope
+  edge @id.lexical_scope -> @this.before_scope
+
+  edge @id.type -> ref__typeof
+
+  attr (ref__typeof) push_symbol = ":"
+  edge ref__typeof -> ref
+
+  edge @this.after_scope -> @this.before_scope
+}
+
+(resource (field_access) @field_access) @this {
+  edge @field_access.lexical_scope -> @this.before_scope
+  edge @this.after_scope -> @this.before_scope
+}
+
+(try_with_resources_statement (catch_clause (catch_formal_parameter (catch_type) @type (_) @id) body: (_) @body)) @stmt {
+  node def
+  node defs
+  attr (def) node_definition = @id
+  edge defs -> def
+  edge @body.before_scope -> defs
+  edge defs -> @stmt.before_scope
+
+  edge @type.lexical_scope -> @stmt.before_scope
+
+  node def__typeof
+  attr (def__typeof) pop_symbol = ":"
+
+  edge def -> def__typeof
+
+  edge def__typeof -> @type.type
+}
+
+(try_with_resources_statement (finally_clause (_) @finally)) @stmt {
+  edge @finally.before_scope -> @stmt.before_scope
+}
+
+(while_statement condition: (_) @condition body: (_) @body) @stmt {
+  edge @condition.lexical_scope -> @stmt.before_scope
+  edge @body.before_scope -> @stmt.before_scope
+  edge @stmt.after_scope -> @stmt.before_scope
+}
 
-  edge @type.value -> @local_var.after_scope
-  edge @type_identifier.value -> @local_var.after_scope
+(yield_statement (_) @expr) @stmt {
+  edge @expr.lexical_scope -> @stmt.before_scope
+  edge @stmt.after_scope -> @stmt.before_scope
 }
 
 ;; =====================
 ;;  Primary Expressions
 ;; =====================
 
+(array_access (primary_expression) @array (expression) @index) @this {
+  edge @array.lexical_scope -> @this.lexical_scope
+  edge @index.lexical_scope -> @this.lexical_scope
+}
+
+(array_creation_expression type: (_) @type) @this {
+  edge @type.lexical_scope -> @this.lexical_scope
+}
+
+(array_creation_expression (dimensions_expr (_) @expr)) @this {
+  edge @expr.lexical_scope -> @this.lexical_scope
+}
+
+(array_initializer (_) @expr) @this {
+  edge @expr.lexical_scope -> @this.lexical_scope
+}
+
+(class_literal (_) @type) @this {
+  edge @type.lexical_scope -> @this.lexical_scope
+}
+
 (primary_expression/identifier) @ident
 {
-  node ref
-  node ref__typeof
+  node member
+  node implicit_this
+  node implicit_this__typeof
 
-  attr (ref) node_reference = @ident
+  node object__typeof
 
+  attr (implicit_this) symbol_reference = "this", empty_source_span
+  edge implicit_this__typeof -> implicit_this
+  attr (implicit_this__typeof) push_symbol = ":"
+  edge implicit_this -> @ident.lexical_scope
```

Comment:
This is a little bit gross, but consider these two examples:

```java
class Foo {
  void f() {
    int x;
    x;
  }
}
```


```java
class Foo {
  int x;
  void f() {
    x;
  }
}
```

Syntactically, we can't distinguish between the two using a query, so we generate paths for both lexically scoped variables and implicit `this` field accesses and allow precedence to prefer the former.

I think that won't work for globals, tho, so this might need to be amended to place instance methods and variables declared on the class into the lexical scope between the local and global scopes. That would result in simpler paths as well, but I'm not sure exactly what the graph will look like. I suppose it's a bit like how we handle `enum` constants.